guava/src/com/google/common/io/ReaderInputStream.java

```diff
@@ -185,6 +185,76 @@ public int read(byte[] b, int off, int len) throws IOException {
     }
   }
 
+  @Override
+  public int available() throws IOException {
+    if (byteBuffer.hasRemaining()) {
+      // if we have remaining chars in the buffer, return their count
+      return byteBuffer.remaining();
+    } else if (charBuffer.hasRemaining()) {
```

Comment:
I think this also has to check `draining == true`, otherwise `CharSource.wrap("").asByteSource(charset).openStream().available()` would return the size of the empty byte buffer instead of 0.

Additionally you need to account for the situation where `draining == true`, but `byteBuffer.hasRemaining() == false` (which I think is the situation where you are experiencing the OVERFLOW).

The following should work (I hope):
```suggestion
    if (draining) {
      if (byteBuffer.hasRemaining()) {
        // if we have remaining bytes in the buffer, return their count
        return byteBuffer.remaining();
      } else {
        // if buffer has been drained completely, clear it then fall through to encoding
        draining = false;
        Java8Compatibility.clear(byteBuffer);
      }
    }
    
    if (charBuffer.hasRemaining()) {
```

----

Alternatively (and maybe better) you could also change `ReaderInputStream.read(...)` to always update `draining` before returning:
```diff
      // We stay in draining mode until there are no bytes left in the output buffer. Then we go
      // back to encoding/flushing.
      if (draining) {
        totalBytesRead += drain(b, off + totalBytesRead, len - totalBytesRead);
+       if (!byteBuffer.hasRemaining()) {
+         draining = false;
+         Java8Compatibility.clear(byteBuffer);
+       }
        if (totalBytesRead == len || doneFlushing) {
          return (totalBytesRead > 0) ? totalBytesRead : -1;
        }
-       draining = false;
-       Java8Compatibility.clear(byteBuffer);
      }
```

Then you only need to change it like this:
```suggestion
    if (draining) {
      // if we have remaining bytes in the buffer, return their count
      return byteBuffer.remaining();
    } else if (charBuffer.hasRemaining()) {
```